extensions/blocks/send-a-message/whatsapp-button/edit.js

```diff
@@ -0,0 +1,260 @@
+/**
+ * External dependencies
+ */
+import { __, _x } from '@wordpress/i18n';
+import { useEffect, useState } from '@wordpress/element';
+import { withDispatch } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+import {
+	Button,
+	BaseControl,
+	TextControl,
+	TextareaControl,
+	SelectControl,
+	PanelBody,
+	ToolbarGroup,
+	Dropdown,
+	Path,
+} from '@wordpress/components';
+import {
+	BlockControls,
+	InspectorControls,
+	RichText,
+	PanelColorSettings,
+} from '@wordpress/block-editor';
+import { DOWN } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import { countryCodes } from '../shared/countrycodes.js';
+import renderMaterialIcon from '../../../shared/render-material-icon';
+import HelpMessage from '../../../shared/help-message';
+import './view.scss';
+
+const WHATSAPP_GREEN = '#25D366';
+const WHATSAPP_DARK = '#465B64';
+const WHATSAPP_LIGHT = '#F4F4F4';
+
+const WhatsAppButtonEdit = ( { attributes, setAttributes, className, selectBlock, clientId } ) => {
+	const {
+		countryCode,
+		phoneNumber,
+		buttonText,
+		firstMessage,
+		colorClass,
+		backgroundColor,
+	} = attributes;
+
+	const [ isValidPhoneNumber, setIsValidPhoneNumber ] = useState( true );
+
+	const getCountryCode = async () => {
+		setAttributes( { countryCode: '1' } );
+
+		const geoFetch = await fetch( 'http://ip-api.com/json/?fields=countryCode' )
+			.then( response => {
+				if ( ! response.ok ) {
+					return false;
+				}
+
+				return response;
+			} )
+			.catch( () => {
+				return false;
+			} );
+
+		if ( geoFetch ) {
+			const geo = await geoFetch.json();
+
+			countryCodes.forEach( item => {
+				if ( item.code === geo.countryCode ) {
+					setAttributes( { countryCode: item.value } );
+				}
+			} );
+		}
+	};
+
+	useEffect( () => {
+		if ( undefined === countryCode ) {
+			getCountryCode();
+			selectBlock( clientId );
+		}
+	} );
+
+	const validatePhoneNumber = newPhoneNumber => {
+		const phoneNumberRegEx = RegExp( /^[+]?[\s./0-9]*[(]?[0-9]{1,4}[)]?[-\s./0-9]*$/, 'g' );
```

Comment:
I'm very bad at regexes, do you mind adding a comment to explaining what we're testing for here? 😄 

Like, what kind of characters do we want to allow?
No alphabetical chars: 👍 
No plus sign at the beginning: 👍 

But then we allow the plus sign in the middle, any number of periods and other special chars.
Eventually, in the `save`, we're going to remove all non-numeric chars anyway (`phoneNumber.replace( /\D+/g )`), maybe we could be more consistent.

Reply:
I'll add a comment to the regex. :+1: The reason for not unifying this is more for user experience. WhatsApp wants a plan number with no decorative characters like dashes, brackets, or dots. For the user this is not as readable, so I want to leave those in when displaying in the input field.

Reply:
Makes sense.
Thanks for adding the comment, very appreciated. 🙇